File: src/mtgsim/api/models/card.py
"""Card-related Pydantic models."""

from pydantic import BaseModel

from mtgsim.api.models.common import CollectionDetail, Pagination


class CardSummary(BaseModel):
    """Summary card information for list views."""

    uuid: str
    name: str
    type: str | None = None
    mana_cost: str | None = None
    mana_value: float | None = None
    rarity: str | None = None
    set_code: str | None = None
    color_identity: list[str] = []
    tags: list[str] = []
    text: str | None = None
    price: float | None = None
    image_url: str | None = None
    owns: bool = False
    wants: bool = False
    total_owned: int = 0
    total_wanted: int = 0


class CardListResponse(BaseModel):
    """Response for card list endpoint."""

    data: list[CardSummary]
    pagination: Pagination


class CardPriceEntry(BaseModel):
    """Single price entry from a provider."""

    provider: str
    finish: str
    listing_type: str
    price: float


class CardAppearance(BaseModel):
    """Deck appearance for a card."""

    file: str
    name: str
    count: int


class CardPrinting(BaseModel):
    """Other printing of a card."""

    set_code: str
    set_name: str
    uuid: str
    rarity: str | None = None
    number: str | None = None
    image_url: str | None = None
    price: float | None = None
    owns: bool = False
    total_owned: int = 0


class QuadrantRating(BaseModel):
    """Quadrant theory rating for a card."""

    developing: float | None = None
    ahead: float | None = None
    behind: float | None = None
    parity: float | None = None
    notes: str | None = None


class CardDetail(BaseModel):
    """Full card details."""

    uuid: str
    name: str
    mana_cost: str | None = None
    mana_value: float | None = None
    type: str | None = None
    types: list[str] = []
    subtypes: list[str] = []
    supertypes: list[str] = []
    text: str | None = None
    flavor_text: str | None = None
    rarity: str | None = None
    set_code: str | None = None
    set_name: str | None = None
    color_identity: list[str] = []
    colors: list[str] = []
    keywords: list[str] = []
    tags: list[str] = []
    power: str | None = None
    toughness: str | None = None
    loyalty: str | None = None
    defense: str | None = None
    artist: str | None = None
    number: str | None = None
    layout: str | None = None
    finishes: list[str] = []
    border_color: str | None = None
    frame_version: str | None = None
    is_reprint: bool = False
    is_reserved: bool = False
    is_promo: bool = False
    image_url: str | None = None
    legalities: dict[str, str] = {}
    all_prices: list[CardPriceEntry] = []
    appears_in_decks: list[CardAppearance] = []
    other_printings: list[CardPrinting] = []
    owns: bool = False
    wants: bool = False
    total_owned: int = 0
    total_wanted: int = 0
    collection: CollectionDetail | None = None
    quadrant_rating: QuadrantRating | None = None

File: frontend/src/types/card.ts
import type { Pagination } from "./common"

export interface CardSummary {
  uuid: string
  name: string
  type: string
  mana_cost: string | null
  mana_value: number
  rarity: string
  set_code: string
  color_identity: string[]
  tags: string[]
  text: string | null
  price: number | null
  image_url: string | null
  owns: boolean
  wants: boolean
  total_owned: number
  total_wanted: number
}

export interface CardListResponse {
  data: CardSummary[]
  pagination: Pagination
}

export interface CardPriceEntry {
  provider: string
  finish: string
  listing_type: string
  price: number
}

export interface CardPrinting {
  set_code: string
  set_name: string
  uuid: string
  rarity: string
  number: string
  image_url: string | null
  owns: boolean
  total_owned: number
}

export interface CardAppearance {
  file: string
  name: string
  count: number
}

export interface CollectionInfo {
  quantity_owned: number
  quantity_owned_foil: number
  quantity_wanted: number
  quantity_wanted_foil: number
  condition: string | null
  notes: string | null
}

export interface QuadrantRating {
  developing: number | null
  ahead: number | null
  behind: number | null
  parity: number | null
  notes: string | null
}

export interface CardDetail {
  uuid: string
  name: string
  mana_cost: string | null
  mana_value: number
  type: string
  types: string[]
  subtypes: string[]
  supertypes: string[]
  text: string | null
  flavor_text: string | null
  rarity: string
  set_code: string
  set_name: string
  color_identity: string[]
  colors: string[]
  keywords: string[]
  tags: string[]
  power: string | null
  toughness: string | null
  loyalty: string | null
  defense: string | null
  artist: string | null
  number: string | null
  layout: string | null
  finishes: string[]
  border_color: string | null
  frame_version: string | null
  is_reprint: boolean
  is_reserved: boolean
  is_promo: boolean
  image_url: string | null
  legalities: Record<string, string>
  all_prices: CardPriceEntry[]
  appears_in_decks: CardAppearance[]
  other_printings: CardPrinting[]
  owns: boolean
  wants: boolean
  total_owned: number
  total_wanted: number
  collection: CollectionInfo | null
  quadrant_rating: QuadrantRating | null
}
